### PR TITLE
Fixing Layer / Neuron Attributions with In-place Operations

### PR DESCRIPTION
**Pull request by vivekmig**
This PR fixes neuron / layer attributions with in-place operations by keeping appropriate clones of intermediate values to ensure that they are not modified by future operations.

Addresses Issue: #156

### CODE REVIEW

**Comment by NarineK** [comment]
Nice! Can we maybe create a similar test model from https://github.com/pytorch/captum/issues/156
with `nn.Sequential` and access the first layer from the module like in the issue ?

**Comment by vivekmig** [comment]
Sure, I've added the test!

**Comment by vivekmig** [comment]
Yup, the backward hook issue only affects the last layer / module before the hook (only for layer output). An example and test for this case is test_deconv_matching, which checks that NeuronDeconvolution matches Deconvolution output, which is not true if the hook doesn't execute on the last layer (which is a ReLU) in the neuron gradient computation.

The two functions do compute different things, but both can compute neuron gradients when given gradient_neuron_index. Since compute_layer_gradients_and_eval computes the layer gradients as well, it needs to hook and clone a little differently (overriding the module output with the clone) which inhibits the last layer's module hook. So if this is necessary, _forward_layer_eval_with_neuron_grads should be used to obtain neuron gradients. Sorry, this probably wasn't very clear in the comment, can update to make it clearer!

File: captum/attr/_utils/gradient.py [hunk 1]
```diff
@@ -175,9 +175,12 @@ def forward_hook(module, inp, out):
...
             # doesn't support a tuple of inputs
-            saved_layer[eval_tsr.device] = eval_tsr
+            saved_layer[eval_tsr.device] = eval_tsr.clone()
 
-    hook = layer.register_forward_hook(forward_hook)
+    if attribute_to_layer_input:
+        hook = layer.register_forward_pre_hook(forward_hook)
+    else:
+        hook = layer.register_forward_hook(forward_hook)
     _run_forward(forward_fn, inputs, additional_forward_args=additional_forward_args)
```
Comment (NarineK) [inline captum/attr/_utils/gradient.py]: I was just looking into this. Doesn't `_run_forward ` also need `target_ind` to be passed as an argument ?

**Pull request closed without merging**